File: 백준/Gold/2294. 동전 2/동전 2.py
import sys

N, K = map(int, sys.stdin.readline().split())

coin = []
dp = [0 for _ in range(K+1)]

for _ in range(N):
    coin.append(int(sys.stdin.readline()))

for i in range(1, K+1):
    temp = []
    for c in coin:
        if c <= i and dp[i-c] != -1:
            temp.append(dp[i-c])
    if not temp:
        dp[i] = -1
    else:
        dp[i] = min(temp) + 1

print(dp[K])

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        int[] coin = new int[N];

        for(int i = 0; i < N; i++){
            coin[i] = Integer.parseInt(br.readLine());
        }

        int[] dp = new int[K + 1];

        Arrays.fill(dp, 987654321);

        for(int i = 0; i < N; i++){
            if(K >= coin[i]){
                dp[coin[i]] = 1;
                for(int j = 0; j <= K; j++){
                    if(j >= coin[i])
                        dp[j] = Math.min(dp[j - coin[i]] + 1, dp[j]);
                }
            }
        }

        System.out.println(dp[K] > 10000 ? -1 : dp[K]);
    }
}
